### PR TITLE
feat(web): improve range selection

### PR DESCRIPTION
**Pull request by brighteyed**
## Description

This PR fixes some range selection UI issues and makes it possible to Shift-select ranges across _all the timeline_

## How Has This Been Tested?

- [x] Select/deselect a single asset
- [x] Select/deselect multiple assets one by one
- [x] Select/deselect date groups
- [x] Select/deselect ranges within one day
- [x] Select/deselect ranges across adjacent days
- [x] Select/deselect ranges across adjacent buckets
- [x] Select/deselect ranges across non-adjacent buckets
- [x] Select/deselect ranges across buckets, including not loaded buckets
- [x] Fuzzy single/multiple selections across all the timeline
- [x] Main timeline scroll up/down

## Demo:
https://github.com/immich-app/immich/assets/71479/8af43640-73b8-4ef1-97ed-126647d5daf6

### CODE REVIEW

**Review by jrasm91 — APPROVED** [review]

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 1]
```diff
@@ -158,18 +157,82 @@
     lastAssetMouseEvent = asset;
   };
 
+  const handleSelectAssets = async (e: CustomEvent) => {
+    const asset = e.detail.asset;
+    if (asset) {
```
Comment (uhthomas) [inline web/src/lib/components/photos-page/asset-grid.svelte:162]: Can we remove a level of indentation by returning early?

```ts
if (!asset) {
  return;
}
```

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 2]
```diff
@@ -158,18 +157,82 @@
...
+            if (deselect) {
+              assetInteractionStore.removeAssetFromMultiselectGroup(asset);
+            } else {
+              assetInteractionStore.addAssetToMultiselectGroup(asset);
+            }
+          }
+        }
+
+        // Update date group selection
+        for (let bucketIndex = startBucketIndex; bucketIndex <= endBucketIndex; ++bucketIndex) {
```
Comment (uhthomas) [inline web/src/lib/components/photos-page/asset-grid.svelte:203]: Definitely `++bucketIndex`? It's unconventional to write loops like this.

File: web/src/lib/components/photos-page/asset-grid.svelte [hunk 3]
```diff
@@ -158,18 +157,82 @@
...
+  };
+
   const selectAssetCandidates = (asset: AssetResponseDto) => {
     if (!shiftKeyIsDown) {
       return;
     }
 
-    const lastSelectedAsset = getLastSelectedAsset();
-    if (!lastSelectedAsset) {
+    const rangeBegin = $assetSelectionStart;
```
Comment (uhthomas) [inline web/src/lib/components/photos-page/asset-grid.svelte:229]: nit: `rangeStart`?

File: web/src/lib/components/assets/thumbnail/thumbnail.svelte
Comment (uhthomas) [inline web/src/lib/components/assets/thumbnail/thumbnail.svelte:1]: Could we also use a fade transition for the button? Something like:

```svelte
class:opacity-0={!readonly && (mouseOver || selected || selectionCandidate)}
```

Comment (brighteyed) [reply web/src/lib/components/assets/thumbnail/thumbnail.svelte:1]: It looks a bit smoother, but with this change, a bunch of focus/tab behavior issues arise. That's definitely not for this PR. I added `focus:outline-none` for now

**Pull request merged**